### ISSUE TITLE
Redundant entries added to .gitignore when user selects "Ignore" in the Changes tab menu

### ISSUE THREAD

**Issue by justinTM** [desktop/desktop#2537]
### Description

From the Changes tab, if the user selects an item (file, directory, etc.) > right clicks > "Ignore", then a duplicate entry is added to the `.gitignore` file. 

Example:
If the file `.DS_Store` is marked "Ignore" for, say, 100 times, then 100 copies of `.DS_Store` entry will be added to the `.gitignore` file, regardless if it already exists.

### Version

**GitHub Desktop version:** `0.8.0`

**OS version:** OS X El Capitan, version `10.11.6`

### Steps to Reproduce

1. [First step]
1. [Second step]
1. [and so on]

**Expected behavior:** Ignore the ignore; check if the file already exists. If it does already exist, ignore the user action (silently or with a message to the user).

**Actual behavior:** Infinite duplicates of the already-existing `.gitignore` entry can be created regardless of existing matching entries.

**Reproduces how often:** Always (not intermittent).

**Comment by shiftkey** [comment]
This behaviour occurs due to the fact that Git will only apply gitignore rules to untracked files - once a file is tracked it doesn't care about your .gitignore file contents.

One way to ensure that you can't get into this loop would be to discard the changes to the tracked file as part of this operation, but that doesn't address the underlying issue about how to handle modifications to already-tracked files.

**Comment by shiftkey** [comment]
@justinTM can you confirm that the file you were trying to ignore is already tracked in Git?

**Comment by justinTM** [comment]
> @justinTM can you confirm that the file you were trying to ignore is already tracked in Git?

Yes, the file `.DS_Store` was tracked (and committed) before the entry in `.gitignore` was created; my bad!

I would suggest the issue still remain open, as there could be some improvement to this functionality, regardless of my brain fart? :)

**Comment by shiftkey** [comment]
@justinTM all good, just adding some context as to what Git is doing here

**Comment by shiftkey** [comment]
Looking into what we'd need to change here:

 - `onIgnore` only takes in a string representing the new pattern [here](https://github.com/desktop/desktop/blob/master/app/src/ui/changes/changed-file.tsx#L110) and [here](https://github.com/desktop/desktop/blob/master/app/src/ui/changes/changed-file.tsx#L120) - we'd need to pass information about the file back to the dispatcher.
 - `onIgnore` also supports wildcard expressions - I'd probably leave that behaviour alone here because going through the repository to remove all track files matching some extension feels like overreach.
 - [Dispatcher.ignore()](https://github.com/desktop/desktop/blob/master/app/src/lib/dispatcher/dispatcher.ts?utf8=%E2%9C%93#L560) would also need to know about the file, so it can pass this through to the `GitStore`
 - [AppStore._ignore()](https://github.com/desktop/desktop/blob/master/app/src/lib/dispatcher/app-store.ts#L2291-L2296) could look at the [AppFileStatus](https://github.com/desktop/desktop/blob/master/app/src/models/status.ts#L34-L41) of the changed file and discard the change (optionally?) if it's a modified file

**Comment by shiftkey** [comment]
If someone wants to take a shot at this, I've added some guidance above.

**Comment by shiftkey** [comment]
Dropping this from the milestone as the work requires some additional thought around UX - see https://github.com/desktop/desktop/pull/3164#issuecomment-339789343 and replies

**Comment by JeanCollas** [comment]
Thanks, I am happy that you already know the issue, I hope it will be fixed soon. If you don't know what to do about the ignored file if it already exists, why not just "ask the user about it"? 
Or provide a way to ignore+delete from the repository.

**Comment by bartclaeys** [comment]
Seems like this issue hasn't been fixed yet?

**Comment by phantomdesvin** [comment]
Hi. It would be ideal that, when the user adds something new to ignore, if there are tracked files that match that ignore command, GitHub Desktop could bring a message like “The following files are already tracked so they won’t be ignored. Select the files that you wan’t to untrack so they can be ignored from now on” and some visual interface for untracking those files. That would keep users from making this mistake and would improve the workflow as right now you have to use the command prompt to untrack files.

**Comment by gabessdsp** [comment]
Just wondering if there are any updates to this?

**Comment by WaGi-Coding** [comment]
Pretty annoying issue, hope it gets resolved in the next few decades already.

**Comment by Birgerev** [comment]
Three year old issue, any fix planned? This thing has been annoying me since the day I started using the desktop app, is there any workaround or temporary fix?

**Comment by justinTM** [comment]
It appeared to me as if Brendan implied it was an error between chair and screen (i.e., NotABug/WontFix). Good luck.

**Comment by vaidasma** [comment]
Bump! seems this has been an issue since 2017, is this still being investigated?

**Comment by lfkdev** [comment]
Still a big problem.

**Comment by steveward** [comment]
Reported in https://github.com/desktop/desktop/issues/19176.

**Comment by sergiou87** [comment]
Reported in #19213

**Comment by DominiqueMakowski** [comment]
bump

**Comment by judasbenhur** [comment]
whats the proper way to handle this case without manually going through committed files which are supposed to be ignored?

**Comment by differenceclouds** [comment]
Ignoring and untracking a file (like [Issue #14923](https://github.com/desktop/desktop/issues/14923) suggests) with one command would be great.

**Comment by JNewhart658** [comment]
agreed it sucks that almost a decade later and no movement on this
